File: frontend/lib/models/user_model.dart
class UserModel {
  const UserModel({
    required this.id,
    required this.fullName,
    this.email,
    this.phoneNumber,
    this.profilePicture,
    this.isActive = true,
    this.roleId,
    this.createdAt,
  });

  final int id;
  final String? fullName;
  final String? email;
  final String? phoneNumber;
  final String? profilePicture;
  final bool isActive;
  final int? roleId;
  final DateTime? createdAt;

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['id'] as int,
      fullName: json['full_name'] as String?,
      email: json['email'] as String?,
      phoneNumber: json['phone_number'] as String?,
      profilePicture: json['profile_picture'] as String?,
      isActive: json['is_active'] as bool? ?? true,
      roleId: json['role_id'] as int?,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'].toString())
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'full_name': fullName,
        'email': email,
        'phone_number': phoneNumber,
        'profile_picture': profilePicture,
        'is_active': isActive,
        'role_id': roleId,
        'created_at': createdAt?.toIso8601String(),
      };

  UserModel copyWith({
    String? fullName,
    String? profilePicture,
  }) {
    return UserModel(
      id: id,
      fullName: fullName ?? this.fullName,
      email: email,
      phoneNumber: phoneNumber,
      profilePicture: profilePicture ?? this.profilePicture,
      isActive: isActive,
      roleId: roleId,
      createdAt: createdAt,
    );
  }
}

File: backend/app/models/user_model.py
from sqlalchemy import Column, Integer, String, Boolean, DateTime, ForeignKey
from sqlalchemy.orm import relationship, validates
from datetime import datetime
from backend.app.db.base import Base


class User(Base):
    __tablename__ = "users"

    id = Column(Integer, primary_key=True, index=True)
    full_name = Column(String, nullable=True)
    email = Column(String(100), unique=True, index=True, nullable=True)
    phone_number = Column(String(20), unique=True, index=True, nullable=True)
    google_id = Column(String(128), unique=True, index=True, nullable=True)
    password = Column(String(255), nullable=False)
    profile_picture = Column(String(255), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow)
    role_id = Column(Integer, ForeignKey('roles.id'))
    is_active = Column(Boolean, default=True)
    is_removed = Column(Boolean, default=False)
    role = relationship("Role")
    player_stats = relationship(
        "PlayerStats",
        back_populates="user",
        uselist=False,
        cascade="all, delete-orphan",
    )
    player_activities = relationship(
        "PlayerActivity",
        back_populates="user",
        cascade="all, delete-orphan",
    )

    @validates('email', 'phone_number')
    def validate_email_or_phone(self, key, value):
        if key == 'email':
            if value is None and self.phone_number is None:
                raise ValueError(
                    'At least either email or phone number must be provided'
                )
        elif key == 'phone_number':
            if value is None and self.email is None:
                raise ValueError(
                    'At least either email or phone number must be provided'
                )
        return value
